scripts/release.mts

```diff
@@ -42,22 +35,90 @@ interface NpmPack {
   files: { path: string; size: number; mode: number }[]
 }
 
+/**
+ * Creates a Github release through the Github CLI, if the release doesn't already exist.
+ * @param packDir The directory where the tarball is located
+ * @param filename The filename of the tarball
+ * @param gitTag The git tag to create the release under
+ * @param changelogPath The path to the changelog
+ */
 async function createGithubRelease(
   packDir: string,
   filename: string,
   gitTag: string,
   changelogPath: string,
 ) {
+  if (await checkIfReleaseExists(gitTag)) {
+    warn(`Release under tag "${gitTag}" already exists, skipping...`)
+    return
+  }
+
   const asset = `${packDir}/${filename}`
   log(
     `Creating release with tag "${gitTag}" with asset "${asset}" and changelog of ${changelogPath}`,
   )
   await $`gh release create ${gitTag} ${packDir}/${filename} -F ${changelogPath}`
+
   log(`Cleaning up...`)
   log(`Removing ${packDir}`)
   fs.remove(packDir)
 }
 
+/**
+ * Modifies the CHANGELOG.md to exclude older version notes so they're not included in the release.
+ * This code is dependent on how the changelog is formatted, and will no-op if the changelog doesn't
+ * match the expected format.
+ * @param changelogPath The path to the changelog
+ * @param gitTag The git tag to include changes for
+ */
+async function modifyChangelog(changelogPath: string, gitTag: string) {
+  log(`Modifying changelog to only include changes for ${gitTag}...`)
+  const changelog = await fs.readFile(changelogPath, 'utf-8')
+
+  // git tag version to changelog version header format (v0.0.0-0 -> ## 0.0.0-0)
+  const version = `## ${gitTag.substring(1)}`
+  const changelogHeaderVersionPattern = /^## [0-9]+.[0-9]+.[0-9]+-?[0-9a-z]*$/gm
```

Comment:
it looks like sonarqube is complaining about this as a risk:
```Make sure the regex used here, which is vulnerable to super-linear runtime due to backtracking, cannot lead to denial of service.```

Reply:
I've modified the regex to try and exclude certain quantifiers like `+` and `*`. I've still included a `?` which might still trigger the check.

Given that this is only for the release aspect this can't really cause a ReDoS. So should be ok to push through - thoughts?

I don't know if this will trigger on other PRs which don't touch this line though.